aio: reimplement nav items

petebacondarwin opened the pull request:
This replaces #14924

petebacondarwin committed:
fix(aio): ensure that only one request for navigation.json is made

petebacondarwin committed:
fix(aio): ensure that only one request per document is made

petebacondarwin committed:
refactor(aio): move document path computation

Do the path computation before doing the cache look up.

petebacondarwin committed:
fix(aio): map the empty url to the correct doc path

petebacondarwin committed:
fix(aio): emit a hard-coded doc if file-not-found fails

Closes #14848

petebacondarwin committed:
docs(aio): make home doc the default home page

Now the `/` url will display the home page.

Closes #14849

petebacondarwin committed:
fix(aio): LocationService urls should never start with a slash

There is a weirdness in the Angular Location service.
If the `baseHref` is only a single slash (`'/'`) then it
changes it to be an empty string (`''`). The effect of this
is that `Location.normaliseUrl(url)` does not strip off the
leading slash from url paths.

The problem is that the leading slash only appears on the
initial Location path, and not on urls that arrive from subscribing
to the Location service.

This commit is a workaround this problem.

petebacondarwin committed:
fix(aio): no need for path and url properties in nav

petebacondarwin committed:
fix(aio): correctly intercept links with LinkDirective

petebacondarwin committed:
fix(aio): move interfaces into their own files to workaround compile weirdness

For some reason the tree-shaker is not picking up these interfaces
(perhaps TS is not passing it through) when they are in the same file
as their related services. This results in a distracting warning message.

petebacondarwin committed:
fix(aio): reimplement side-nav

wardbell commented on aio/src/app/layout/nav-item/nav-item.component.scss:
```diff
@@ -44,71 +44,59 @@ To use these, put this snippet in the approriate selector:
+    text-transform: uppercase;
+  }
 
-.heading.selected.level-1,
```
I found the problem. See comment on removing a css rule

wardbell commented on aio/src/app/layout/nav-item/nav-item.component.scss:
```diff
@@ -44,71 +44,59 @@ To use these, put this snippet in the approriate selector:
-  display: block;
-}
+    &.level-1,
+    &.level-1 .children {
```
This line is causing 3 bars to appear on a 3-deep selection. We only want ONE bar on the left side, running the length of the selected top level nodes.  Remove `&.level-1 .children` and I think we're back to a good state.

petebacondarwin replied:
Yep, I misunderstood the original CSS.

wardbell commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -1,23 +1,19 @@
 import { combineLatest } from 'rxjs/observable/combineLatest';
-import 'rxjs/add/operator/publish';
+import 'rxjs/add/operator/publishReplay';
+import 'rxjs/add/operator/publishLast';
```
`publishLast` is being deprecated, last I heard.

petebacondarwin replied:
where did you hear that? There is no mention of it in the docs or the changelog or the source code of rxjs.

wardbell replied:
I know. Probably part of the noise around `publishReplay`.  Not to worry for now.

wardbell commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -34,41 +30,47 @@ const NAVIGATION_PATH = 'content/navigation.json';
-    this.currentNode = currentMapItem.map(item => item.node).publish();
-    this.activeNodes = currentMapItem.map(item => [item.node, ...item.parents]).publish();
+  private fetchNavigation(): Observable<NavigationViews> {
+    const response = this.http.get(NAVIGATION_PATH)
```
Per style guide, constants are no longer ALL_CAPS. They are are in lower camelCase. Should be like `navigationPath`.

wardbell replied:
Doesn't return an observable of response but rather an observable of `NavigationViews`. Variable should be something like `navViews`

petebacondarwin replied:
done

petebacondarwin replied:
done

wardbell commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -34,41 +30,47 @@ const NAVIGATION_PATH = 'content/navigation.json';
-    return this.http.get(NAVIGATION_PATH).map(res => res.json() as NavigationViews);
+  private getActiveNodes() {
+    const currentMapItem = combineLatest(
+      this.navigationViews.map(this.computeNavMap),
```
`computeNavMap` belongs on the `navigationViews` observable, not here. We don't really have navigationViews until this function runs. 

Moreover, the result is a NavMap. So really want an observable of NavMap, not of NavigationViews. 

I'm really confused why we are exposing both NavigationViews and a NavMap. Can't we just have one?

petebacondarwin replied:
I think you are confused.
The `navigationViews` observable is a representation of the tree of navigation nodes as loaded up from the JSON requested from the server. This is used to create the navigational UI (i.e. the nav-items, etc).

The navMap is an internal structure that maps a URL to a node (and its ancestors) in the navigationViews tree.  The `selectedNodes` observable respresents the node (and its ancestors) that map to the current URL. The map itself is not exposed outside of the `NavigationService`.

The way this is built either the current URL can change or the actual navigationViews (via a refresh maybe :-P ), will cause the list of selected nodes to update.

Creating the navigation tree view and indicating which nodes are selected are separate features.

I can see that we can optimise the structure and remove the distinction between the `node` and the `parents` (ancestors) properties in the `NavigationMapItem` so I will do that, which will simplify the code further.

wardbell commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -34,41 +30,47 @@ const NAVIGATION_PATH = 'content/navigation.json';
-  private fetchNavigation(): Observable<NavigationViews> {
-    // TODO: logging and error handling
-    return this.http.get(NAVIGATION_PATH).map(res => res.json() as NavigationViews);
+  private getActiveNodes() {
```
TBH, this does not feel like an improvement to me. I don't know why the NavigationService creates a separate stream of ActiveNodes when the only thing that cares about that is the navigation item which can decide if a node is active by looking at if it's among the ancestors.

That's what you're doing all over again anyway in the NavItem _anyway_, albeit with a slightly different test.

For my money ... this service - this method -  is much more complicated than it needs to be with no compensating savings elsewhere.  Someone reading it a week from now is going to wonder what it does. There are no comments. This is going to be a perpetual source of mystery.

What was wrong with the original?

petebacondarwin replied:
See my comment further down

wardbell commented on aio/src/app/layout/nav-item/nav-item.component.ts:
```diff
@@ -1,50 +1,58 @@
-          target={{node.target}}
-          title={{node.title}}
-          class="vertical-menu">
+    <div [ngClass]="levelClass" [class.active]="active" [class.heading]="node.children" class="nav-item">
```
I would prefer a separate template rather than an inline template.

petebacondarwin replied:
done

wardbell commented on aio/src/app/layout/nav-item/nav-item.component.ts:
```diff
@@ -1,50 +1,58 @@
-      active: false,
-      heading: !!this.node.children
-    };
+    this.activeNodesSubscription = this.nav.activeNodes.subscribe(activeNodes => this.checkActiveNodes(activeNodes));
```
I don't love your decision to make every single item subscribe and unsubscribe. 

I preferred the original in which a container view subscribed and communicated to the more numerous and simple nav-items were simpler through a simple input binding.

I don't know that it matters but it threw me off.

petebacondarwin replied:
I changed it back to have the information being passed down the tree.
I can see that it is nice to have zero observable cruft in a component like this. Although we do now get lots of change detection updates instead.

wardbell reviewed (changes requested):
I left some notes.

The crucial fixes are:
1. CSS so that we don't have 3 bars and confusing selection.
1. Restore the technique for moving the top nav to the left nave when in "phone mode"
1. Restore the animations.

petebacondarwin commented on aio/src/app/layout/nav-item/nav-item.component.scss:
```diff
@@ -44,71 +44,59 @@ To use these, put this snippet in the approriate selector:
-  top: 6px;
-  // left: 4px;
-}
+  &.active {
```
except that in the `NavigationService` we have a property called `activeNodes`

petebacondarwin commented on aio/src/app/layout/nav-item/nav-item.component.ts:
```diff
@@ -1,50 +1,58 @@
-          <md-icon [class.active]="!isActive">keyboard_arrow_right</md-icon>
-          <md-icon [class.active]="isActive">keyboard_arrow_down</md-icon>
-        </ng-template>
+        <md-icon *ngIf="node.children && !expanded">keyboard_arrow_right</md-icon>
```
where were the animations defined?

petebacondarwin replied:
The code that I "deleted" did not have a rotating `>` it had these two `md-icon`s.
A rotating icon would be cool. Perhaps you have it on another branch somewhere and can provide a PR?

petebacondarwin committed:
refactor(aio): rename `activeNodes` to `selectedNodes`

This keeps the naming consistent with the nav-menu
and nav-item component properties

petebacondarwin committed:
docs(aio): copy devguide images from angular.io

petebacondarwin committed:
docs(aio): fix paths to images

petebacondarwin commented on aio/src/app/mocks/logger.service.ts:
```diff
@@ -0,0 +1,23 @@
+import { Injectable } from '@angular/core';
```
I am happy to move mocks into a `testing` folder.

petebacondarwin replied:
done

petebacondarwin committed:
test(aio): add mock services

petebacondarwin committed:
style(aio): rename const to camelCase

petebacondarwin committed:
style(aio): rename local variable

petebacondarwin commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -34,41 +30,47 @@ const NAVIGATION_PATH = 'content/navigation.json';
-        // only map to this item if it has a doc associated with it
-        navMap[node.path] = item;
+    function walkNodes(node: NavigationNode, parent: NavigationMapItem | null) {
+      const item: NavigationMapItem = { node, parents: [] };
```
changed

petebacondarwin commented on aio/src/app/layout/nav-item/nav-item.component.ts:
```diff
@@ -1,50 +1,58 @@
-          title={{node.title}}
-          class="vertical-menu">
+    <div [ngClass]="levelClass" [class.active]="active" [class.heading]="node.children" class="nav-item">
+      <a *ngIf="node.url" href="{{node.url}}" target="{{node.target}}" title="{{node.title}}" (click)="toggleExpanded()">
```
we have the `LinkDirective` that manages the navigation when the user clicks.

petebacondarwin committed:
refactor(aio): simplify and document the NavigationService

petebacondarwin committed:
fix(aio): reimplement the top menu dropping into the side bar

I couldn't find the CSS for the `.small` class to make it look nice.

petebacondarwin commented:
@wardbell I believe I have dealt with all your concerns - one way or another...

In particular these crucial issues:
1. CSS so that we don't have 3 bars and confusing selection.
   **done**
2. Restore the technique for moving the top nav to the left nave when in "phone mode"
  **done - except perhaps for some styling issues**
3. Restore the animations.
  **I couldn't find any animation stuff. Perhaps you can point me at it?**

wardbell commented:
I'll add the animations back in a follow-on PR. Forward!
